File: python-server/storage/file_storage.py
"""File-based JSON persistence for cells."""
import os
import json
from typing import List, Optional


DATA_DIR = os.environ.get("DASHAWS_DATA_DIR", os.path.join(os.getcwd(), "data-python"))
CELLS_FILE = os.path.join(DATA_DIR, "cells.json")


def _ensure_dir():
    os.makedirs(DATA_DIR, exist_ok=True)


def _read_cells() -> List[dict]:
    try:
        if not os.path.exists(CELLS_FILE):
            return []
        with open(CELLS_FILE, "r", encoding="utf-8") as f:
            cells = json.load(f)
        # Migrate legacy cells without language field
        for cell in cells:
            if "language" not in cell:
                cell["language"] = "python"
        return cells
    except (json.JSONDecodeError, IOError):
        return []


def _write_cells(cells: List[dict]):
    _ensure_dir()
    with open(CELLS_FILE, "w", encoding="utf-8") as f:
        json.dump(cells, f, indent=2, ensure_ascii=False)


class FileStorageBackend:
    async def list(self) -> List[dict]:
        return _read_cells()

    async def get(self, id: str) -> Optional[dict]:
        cells = _read_cells()
        for c in cells:
            if c.get("id") == id:
                return c
        return None

    async def save(self, cell: dict):
        cells = _read_cells()
        cell["updatedAt"] = int(__import__("time").time() * 1000)
        idx = None
        for i, c in enumerate(cells):
            if c.get("id") == cell.get("id"):
                idx = i
                break
        if idx is not None:
            cells[idx] = cell
        else:
            cells.append(cell)
        _write_cells(cells)

    async def delete(self, id: str):
        cells = _read_cells()
        _write_cells([c for c in cells if c.get("id") != id])

File: server/storage/file-storage.ts
import type { Cell, StorageBackend } from '../../src/types/cell.js';
import { readFileSync, writeFileSync, existsSync, mkdirSync } from 'fs';
import { join } from 'path';

const DATA_DIR = process.env.SCRIPT_DASHBOARD_DATA_DIR || join(process.cwd(), 'data');
const CELLS_FILE = join(DATA_DIR, 'cells.json');

function ensureDir(): void {
  if (!existsSync(DATA_DIR)) mkdirSync(DATA_DIR, { recursive: true });
}

function readCells(): Cell[] {
  try {
    if (!existsSync(CELLS_FILE)) return [];
    return JSON.parse(readFileSync(CELLS_FILE, 'utf-8'));
  } catch {
    return [];
  }
}

function writeCells(cells: Cell[]): void {
  ensureDir();
  writeFileSync(CELLS_FILE, JSON.stringify(cells, null, 2));
}

export class FileStorageBackend implements StorageBackend {
  async list(): Promise<Cell[]> {
    return readCells();
  }

  async get(id: string): Promise<Cell | null> {
    const cells = readCells();
    return cells.find(c => c.id === id) ?? null;
  }

  async save(cell: Cell): Promise<void> {
    const cells = readCells();
    const idx = cells.findIndex(c => c.id === cell.id);
    if (idx >= 0) {
      cells[idx] = { ...cell, updatedAt: Date.now() };
    } else {
      cells.push({ ...cell, updatedAt: Date.now() });
    }
    writeCells(cells);
  }

  async delete(id: string): Promise<void> {
    const cells = readCells();
    writeCells(cells.filter(c => c.id !== id));
  }
}
